README.md
[![Build Status](https://travis-ci.org/karupanerura/p5-JSON5.svg?branch=master)](https://travis-ci.org/karupanerura/p5-JSON5) [![Coverage Status](http://codecov.io/github/karupanerura/p5-JSON5/coverage.svg?branch=master)](https://codecov.io/github/karupanerura/p5-JSON5?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/JSON5.svg)](https://metacpan.org/release/JSON5)
# NAME

JSON5 - The JSON5 implementation for Perl 5

# SYNOPSIS

```perl
use JSON5;

my $object = decode_json5('{$ref:"#"}');
# { '$ref' => "#" }
```

# DESCRIPTION

JSON5 is the JSON5 implementation for Perl 5

# FUNCTIONAL INTERFACE

Some documents are copied and modified from ["FUNCTIONAL INTERFACE" in JSON::PP](https://metacpan.org/pod/JSON::PP#FUNCTIONAL-INTERFACE).

## decode\_json5

```
$perl_scalar = decode_json5 $json_text
```

expects an UTF-8 (binary) string and tries to parse that as
an UTF-8 encoded JSON5 text, returning the resulting reference.

This function call is functionally identical to:

```
$perl_scalar = JSON5->new->utf8->decode($json_text)
```

# METHODS

Some documents are copied and modified from ["METHODS" in JSON5::PP](https://metacpan.org/pod/JSON5::PP#METHODS).

## new

```
$json5 = JSON5->new
```

Returns a new JSON5 object that can be used to decode JSON5
strings.

All boolean flags described below are by default _disabled_.

The mutators for flags all return the JSON5 object again and thus calls can
be chained:

```perl
my $object = JSON5->new->utf8->allow_nonref->decode("true")
=> JSON::PP::true
```

## utf8

```
$json5 = $json5->utf8([$enable])

$enabled = $json5->get_utf8
```

If $enable is true (or missing), then the decode method expects to be handled
an UTF-8-encoded string. Please note that UTF-8-encoded strings do not contain any
characters outside the range 0..255, they are thus useful for bytewise/binary I/O.

(In Perl 5.005, any character outside the range 0..255 does not exist.
See to ["UNICODE HANDLING ON PERLS"](#unicode-handling-on-perls).)

In future versions, enabling this option might enable autodetection of the UTF-16 and UTF-32
encoding families, as described in RFC4627.

If $enable is false, then the decode expects thus a Unicode string. Any decoding
(e.g. to UTF-8 or UTF-16) needs to be done yourself, e.g. using the Encode module.

Example, decode UTF-32LE-encoded JSON5:

```perl
use Encode;
$object = JSON5->new->decode (decode "UTF-32LE", $json5text);
```

## allow\_nonref

```
$json5 = $json5->allow_nonref([$enable])

$enabled = $json5->get_allow_nonref
```

If `$enable` is true (or missing), then the `decode` method will accept a
non-reference into its corresponding string, number or null JSON5 value,
which is an extension to RFC4627.

If `$enable` is false, then the `decode` method will croak if
given something that is not a JSON5 object or array.

## max\_size

```
$json5 = $json5->max_size([$maximum_string_size])

$max_size = $json5->get_max_size
```

Set the maximum length a JSON5 text may have (in bytes) where decoding is
being attempted. The default is `0`, meaning no limit. When `decode`
is called on a string that is longer then this many bytes, it will not
attempt to decode the string but throw an exception.

If no argument is given, the limit check will be deactivated (same as when
`0` is specified).

## decode

```
$perl_scalar = $json5->decode($json5_text)
```

JSON5 numbers and strings become simple Perl scalars. JSON5 arrays become
Perl arrayrefs and JSON5 objects become Perl hashrefs. `true` becomes
`1` (`JSON::PP::true`), `false` becomes `0` (`JSON::PP::false`),
`NaN` becomes `'NaN'`, `Infinity` becomes `'Inf'`, and
`null` becomes `undef`.

# SEE ALSO

[JSON::PP](https://metacpan.org/pod/JSON::PP)

# LICENSE

Copyright (C) karupanerura.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

karupanerura <karupa@cpan.org>


Changes
Revision history for Perl extension JSON5

{{$NEXT}}

0.01 2016-12-30T10:44:41Z

    - original version



lib/JSON5/Parser.pm
package JSON5::Parser;
use strict;
use warnings;
use utf8;

use Carp qw/croak/;
use JSON::PP;
use Encode;

our $ROOT;
our $POINTER;

sub new {
    my $class = shift;
    return bless +{
        utf8             => 0,
        allow_nonref     => 0,
        max_size         => 0,
        inflate_boolean  => sub { $_[0] eq 'true' ? JSON::PP::true : JSON::PP::false },
        inflate_nan      => sub { 0+'NaN' },
        inflate_null     => sub { undef },
        inflate_infinity => sub { $_[0] eq '+' ? 0+'Inf' : 0+'-Inf' },
    } => $class;
}

# define accessors
BEGIN {
    # boolean accessors
    for my $attr (qw/utf8 allow_nonref/) {
        my $attr_accessor = sub {
            my $self = shift;
            $self->{$attr} = @_ ? shift : 1;
            return $self;
        };
        my $attr_getter = sub {
            my $self = shift;
            return $self->{$attr};
        };

        no strict qw/refs/;
        *{"$attr"}     = $attr_accessor;
        *{"get_$attr"} = $attr_getter;
    }

    # value accessors
    for my $attr (qw/max_size inflate_boolean inflate_nan inflate_null inflate_infinity/) {
        my $attr_accessor = sub {
            my $self = shift;
            $self->{$attr} = shift if @_;
            return $self;
        };
        my $attr_getter = sub {
            my $self = shift;
            return $self->{$attr};
        };

        no strict qw/refs/;
        *{"$attr"}     = $attr_accessor;
        *{"get_$attr"} = $attr_getter;
    }
}

sub parse {
    my ($self, $content) = @_;
    if (my $max_size = $self->{max_size}) {
        use bytes;
        my $bytes = length $content;
        $bytes <= $max_size
            or croak sprintf 'attempted decode of JSON5 text of %s bytes size, but max_size is set to %s', $bytes, $max_size;
    }
    if ($self->{utf8}) {
        $content = Encode::decode_utf8($content);
    }

    # normalize linefeed
    $content =~ s!\r\n?!\n!mg;

    local $ROOT;
    local $POINTER = \$ROOT;

    $self->_parse() for $content;

    return $ROOT;
}

sub _parse {
    my $self = shift;

    $self->_parse_value();
    return if m!\G(?:\s*|//.*$|/\*.*?\*/)*\z!msgc;
    $self->_error('Syntax Error');
}

sub _skip_whitespace { /\G\s*/msgc }
sub _skip_comments { m!\G//.*$!mgc || m!\G/\*.*?\*/!msgc }

sub _parse_value {
    my $self = shift;

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    my $allow_nonref = $self->{allow_nonref} || $POINTER != \$ROOT;

    if ($self->_parse_object_or_array()) {
        return 1;
    }
    elsif ($allow_nonref) {
        if ($self->_parse_number()) {
            return 1;
        }
        elsif ($self->_parse_boolean()) {
            return 1;
        }
        elsif ($self->_parse_string()) {
            return 1;
        }
        elsif (/\Gnull/mgc) {
            ${$POINTER} = $self->{inflate_null}->();
            return 1;
        }
    }

    return;
}

sub _parse_object_or_array {
    my $self = shift;

    if (/\G\{/mgc) {
        local $POINTER = ${$POINTER} = {};
        return $self->_parse_object_kv();
    }
    elsif (/\G\[/mgc) {
        local $POINTER = ${$POINTER} = [];
        return $self->_parse_array_value();
    }

    return;
}

sub _parse_object_kv {
    my $self = shift;

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    # is last?
    if (/\G\}/mgc) {
        return 1;
    }

    # parse key
    my $key; {
        local $POINTER = \$key;
        if (!$self->_parse_string() && !$self->_parse_identifier()) {
            return;
        }
    }

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    # parse object key sep
    unless (/\G\:/mgc) {
        return;
    }

    # parse value
    my $value; {
        local $POINTER = \$value;
        if (!$self->_parse_value()) {
            return;
        }
    }

    # set value
    my $hash = $POINTER;
    $hash->{$key} = $value;

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    # is last?
    if (/\G\}/mgc) {
        return 1;
    }
    elsif (/\G,/mgc) {
        return $self->_parse_object_kv;
    }

    return;
}

sub _parse_array_value {
    my $self = shift;

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    # is last?
    if (/\G\]/mgc) {
        return 1;
    }

    # parse value
    my $value; {
        local $POINTER = \$value;
        if (!$self->_parse_value()) {
            return;
        }
    }

    # set value
    my $array = $POINTER;
    push @$array => $value;

    # skip
    1 while $self->_skip_whitespace() || $self->_skip_comments();

    # is last?
    if (/\G\]/mgc) {
        return 1;
    }
    elsif (/\G,/mgc) {
        return $self->_parse_array_value;
    }

    return;
}

sub _parse_number {
    my $self = shift;

    if (/\G([-+])?Infinity/mgc) {
        my $number = $self->{inflate_infinity}->($1 || '+');
        ${$POINTER} = $number;
        return 1;
    }
    elsif (/\GNaN/mgc) {
        my $number = $self->{inflate_nan}->();
        ${$POINTER} = $number;
        return 1;
    }
    elsif (/\G([-+]?)0x([0-9a-f]+)/imgc) {
        my $number = hex $2;
        $number *= -1 if $1 && $1 eq '-';
        ${$POINTER} = $number;
        return 1;
    }
    elsif (/\G([-+]?)([0-9]+)\.e([0-9]+)/mgc) {
        my $number = $2 * 10 ** $3;
        $number *= -1 if $1 && $1 eq '-';
        ${$POINTER} = $number;
        return 1;
    }
    elsif (/\G([-+]?(?:[0-9]+(?:\.[0-9]*)?|[0-9]*\.[0-9]+))/mgc) {
        my $number = 0+$1;
        ${$POINTER} = $number;
        return 1;
    }

    return;
}

sub _parse_boolean {
    my $self = shift;

    if (/\G(true|false)/mgc) {
        my $bool = $self->{inflate_boolean}->($1);
        ${$POINTER} = $bool;
        return 1;
    }

    return;
}

sub _parse_string {
    my $self = shift;

    if (/\G(?:"((?:.|(?<=\\)\n)*?)(?<!(?<!\\)\\)"|\'((?:.|(?<=\\)\n)*?)(?<!(?<!\\)\\)\')/mgc) {
        my $str = join '', grep defined, $1, $2;

        # ignore escaped linefeed
        $str =~ s!\\\n!!xmg;

        # de-escape
        $str =~ s!\\b !\x08!xmg;      # backspace       (U+0008)
        $str =~ s!\\t !\x09!xmg;      # tab             (U+0009)
        $str =~ s!\\n !\x0A!xmg;      # linefeed        (U+000A)
        $str =~ s!\\f !\x0C!xmg;      # form feed       (U+000C)
        $str =~ s!\\r !\x0D!xmg;      # carriage return (U+000D)
        $str =~ s!\\" !\x22!xmg;      # quote           (U+0022)
        $str =~ s!\\' !\x27!xmg;      # single-quote    (U+0027)
        $str =~ s!\\/ !\x2F!xmg;      # slash           (U+002F)
        $str =~ s!\\\\!\x5C!xmg;      # backslash       (U+005C)
        $str =~ s{\\u([0-9A-Fa-f]{4})}{# unicode         (U+XXXX)
            chr hex $1
        }xmge;
        $str =~ s{\\U([0-9A-Fa-f]{8})}{# unicode         (U+XXXXXXXX)
            chr hex $1
        }xmge;

        ${$POINTER} = $str;
        return 1;
    }

    return;
}

sub _parse_identifier {
    my $self = shift;

    if (/\G([a-z_\$][0-9a-z_\$]*)/imgc) {
        my $identifier = $1;
        ${$POINTER} = $identifier;
        return 1;
    }

    return;
}

sub _error {
    my ($self, $msg) = @_;

    my $src   = $_;
    my $line  = 1;
    my $start = pos $src || 0;
    while ($src =~ /$/smgco and pos $src <= pos) {
        $start = pos $src;
        $line++;
    }
    my $end = pos $src;
    my $len = pos() - $start;
    $len-- if $len > 0;

    my $trace = join "\n",
        "${msg}: line:$line",
        substr($src, $start || 0, $end - $start),
        (' ' x $len) . '^';
    die $trace, "\n";
}

1;
__END__


lib/JSON5.pm
package JSON5;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use Exporter 'import';
our @EXPORT = qw/decode_json5/;

use JSON::PP;
use JSON5::Parser;

my $JSON5; # cache

sub decode_json5 { ($JSON5 ||= JSON5->new->utf8)->decode(@_) }

sub new {
    my $class = shift;
    return bless {
        parser => JSON5::Parser->new,
    } => $class;
}

sub true;
sub false;
*true  = \&JSON::PP::true;
*false = \&JSON::PP::false;

# define accessors
BEGIN {
    for my $attr (qw/utf8 allow_nonref max_size inflate_boolean inflate_nan inflate_null inflate_infinity/) {
        my $attr_accessor = sub {
            my $self = shift;
            $self->{parser}->$attr(@_);
            return $self;
        };
        my $attr_getter = sub {
            my $self = shift;
            my $get_attr = "get_$attr";
            return $self->{parser}->$get_attr;
        };

        no strict qw/refs/;
        *{"$attr"}     = $attr_accessor;
        *{"get_$attr"} = $attr_getter;
    }
}

sub decode {
    my $self = shift;
    return $self->{parser}->parse(@_);
}

1;
__END__

=encoding utf-8

=head1 NAME

JSON5 - The JSON5 implementation for Perl 5

=head1 SYNOPSIS

    use JSON5;

    my $object = decode_json5('{$ref:"#"}');
    # { '$ref' => "#" }

=head1 DESCRIPTION

JSON5 is the JSON5 implementation for Perl 5

=head1 FUNCTIONAL INTERFACE

Some documents are copied and modified from L<JSON::PP/FUNCTIONAL INTERFACE>.

=head2 decode_json5

    $perl_scalar = decode_json5 $json_text

expects an UTF-8 (binary) string and tries to parse that as
an UTF-8 encoded JSON5 text, returning the resulting reference.

This function call is functionally identical to:

    $perl_scalar = JSON5->new->utf8->decode($json_text)

=head1 METHODS

Some documents are copied and modified from L<JSON5::PP/METHODS>.

=head2 new

    $json5 = JSON5->new

Returns a new JSON5 object that can be used to decode JSON5
strings.

All boolean flags described below are by default I<disabled>.

The mutators for flags all return the JSON5 object again and thus calls can
be chained:

   my $object = JSON5->new->utf8->allow_nonref->decode("true")
   => JSON::PP::true

=head2 utf8

    $json5 = $json5->utf8([$enable])
    
    $enabled = $json5->get_utf8

If $enable is true (or missing), then the decode method expects to be handled
an UTF-8-encoded string. Please note that UTF-8-encoded strings do not contain any
characters outside the range 0..255, they are thus useful for bytewise/binary I/O.

(In Perl 5.005, any character outside the range 0..255 does not exist.
See to L<UNICODE HANDLING ON PERLS>.)

In future versions, enabling this option might enable autodetection of the UTF-16 and UTF-32
encoding families, as described in RFC4627.

If $enable is false, then the decode expects thus a Unicode string. Any decoding
(e.g. to UTF-8 or UTF-16) needs to be done yourself, e.g. using the Encode module.

Example, decode UTF-32LE-encoded JSON5:

  use Encode;
  $object = JSON5->new->decode (decode "UTF-32LE", $json5text);

=head2 allow_nonref

    $json5 = $json5->allow_nonref([$enable])
    
    $enabled = $json5->get_allow_nonref

If C<$enable> is true (or missing), then the C<decode> method will accept a
non-reference into its corresponding string, number or null JSON5 value,
which is an extension to RFC4627.

If C<$enable> is false, then the C<decode> method will croak if
given something that is not a JSON5 object or array.

=head2 max_size

    $json5 = $json5->max_size([$maximum_string_size])
    
    $max_size = $json5->get_max_size

Set the maximum length a JSON5 text may have (in bytes) where decoding is
being attempted. The default is C<0>, meaning no limit. When C<decode>
is called on a string that is longer then this many bytes, it will not
attempt to decode the string but throw an exception.

If no argument is given, the limit check will be deactivated (same as when
C<0> is specified).

=head2 decode

    $perl_scalar = $json5->decode($json5_text)

JSON5 numbers and strings become simple Perl scalars. JSON5 arrays become
Perl arrayrefs and JSON5 objects become Perl hashrefs. C<true> becomes
C<1> (C<JSON::PP::true>), C<false> becomes C<0> (C<JSON::PP::false>),
C<NaN> becomes C<'NaN'>, C<Infinity> becomes C<'Inf'>, and
C<null> becomes C<undef>.

=head1 SEE ALSO

L<JSON::PP>

=head1 LICENSE

Copyright (C) karupanerura.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

karupanerura E<lt>karupa@cpan.orgE<gt>

=cut

